deployment/keystone/capability_management.go

```diff
@@ -2,69 +2,81 @@ package keystone
 
 import (
 	"fmt"
-	"strings"
 
+	"github.com/smartcontractkit/ccip-owner-contracts/pkg/proposal/timelock"
 	"github.com/smartcontractkit/chainlink-common/pkg/logger"
 	"github.com/smartcontractkit/chainlink/deployment"
 	kcr "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/keystone/generated/capabilities_registry"
 )
 
 // AddCapabilities adds the capabilities to the registry
 // it tries to add all capabilities in one go, if that fails, it falls back to adding them one by one
-func AddCapabilities(lggr logger.Logger, registry *kcr.CapabilitiesRegistry, chain deployment.Chain, capabilities []kcr.CapabilitiesRegistryCapability) error {
+
+func AddCapabilities(lggr logger.Logger, registry *kcr.CapabilitiesRegistry, chain deployment.Chain, capabilities []kcr.CapabilitiesRegistryCapability, useMCMS bool) ([]timelock.MCMSWithTimelockProposal, error) {
 	if len(capabilities) == 0 {
-		return nil
+		return nil, nil
 	}
-	// dedup capabilities
-	var deduped []kcr.CapabilitiesRegistryCapability
-	seen := make(map[string]struct{})
-	for _, cap := range capabilities {
-		if _, ok := seen[CapabilityID(cap)]; !ok {
-			seen[CapabilityID(cap)] = struct{}{}
-			deduped = append(deduped, cap)
-		}
+	deduped, err := dedupCapabilities(registry, capabilities)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dedup capabilities: %w", err)
 	}
-
-	tx, err := registry.AddCapabilities(chain.DeployerKey, deduped)
+	txOpts := chain.DeployerKey
+	if useMCMS {
+		txOpts = deployment.SimTransactOpts()
+	}
+	tx, err := registry.AddCapabilities(txOpts, deduped)
 	if err != nil {
 		err = DecodeErr(kcr.CapabilitiesRegistryABI, err)
-		// no typed errors in the abi, so we have to do string matching
-		// try to add all capabilities in one go, if that fails, fall back to 1-by-1
-		if !strings.Contains(err.Error(), "CapabilityAlreadyExists") {
-			return fmt.Errorf("failed to call AddCapabilities: %w", err)
-		}
-		lggr.Warnw("capabilities already exist, falling back to 1-by-1", "capabilities", deduped)
```

Comment:
hmmm very sad to see this go but I guess with MCMS it doesn't make sense any more? Can't we simply append all the 1-by-1 transactions? 

If not then we will need to be careful in changesets to only include new capabilities...?

Reply:
the code dedupes them upfront rather than fails and retries. so api behavior is the same.

yes, current logic can't work with mcms

Reply:
see dedupedCapabilities func

and note that all existing tests continue to pass (which exercise duplicate adds)